setwd("C:/Users/Guillem/Documents/PhD/comput/wrkng_dirs_clean/diffsFMT")

load("C:/Users/Guillem/Documents/PhD/comput/wrkng_dirs_clean/sequence_analysis/ParsedSubGraphs_allStrains_named_new.RData")
load("C:/Users/Guillem/Documents/PhD/comput/wrkng_dirs_clean/swarmAnalysis/rhamnMat.RData")
load("C:/Users/Guillem/Documents/PhD/comput/wrkng_dirs_clean/genePresAbs/dictEnzymes.RData")
load("C:/Users/Guillem/Documents/PhD/comput/wrkng_dirs_clean/dictionary/dictionary.RData")
load("C:/Users/Guillem/Documents/PhD/comput/wrkng_dirs_clean/figure3/loadsPaths.RData")
load("C:/Users/Guillem/Documents/PhD/comput/wrkng_dirs_clean/sequence_analysis/geneCallsAllStrains4HeronProkka.RData")
load("C:/Users/Guillem/Documents/PhD/comput/wrkng_dirs_clean/genePresAbs/gffsAllStrains.RData")


if(!require(DECIPHER)) BiocManager::install("DECIPHER")
library(DECIPHER)
if(!require(ggplot2)) install.packages("ggplot2")
library(ggplot2)
if(!require(ggdendro)) install.packages("ggdendro")
library(ggdendro)
if(!require(KEGGREST)) install.packages("KEGGREST")
library(KEGGREST)
if(!require(caret)) install.packages("caret")
library(caret)
if(!require(R.utils)) install.packages("R.utils")
library(R.utils)
if(!require(doParallel)) install.packages("doParallel")
library(doParallel)
source("C:/Users/Guillem/Documents/PhD/comput/wrkng_dirs_clean/MSK_repo_new/randForestSNPs_functions.R")


nbCore <- parallel::detectCores() - 1

library(doParallel)

doParallel::registerDoParallel(nbCore)
showConnections()

# Add strains that are not in metabolomics data and change names of references

rhamnMat <- rbind.data.frame(rhamnMat, data.frame(strains = c("M55212", "F23197", "F5677", "PA7", "PAO1"),
                                                  rhamn3cats = c(0, 2, 0, 2, 2),
                                                  rhamn2cats = c(0, 1, 0, 1, 1)))
newNames <- as.character(rhamnMat$strains)
newNames[c(1, 2)] <- c("PA14_jbx", "UCBPP.PA14")
rhamnMat$strains <- as.factor(newNames)

# Build object with all genomes

allSeqs <- list.files(path = "C:/Users/Guillem/Documents/PhD/comput/Sequences/newSeqs/Archive")

allGenomes <- list()
for(i in seq_along(allSeqs)){
        genome <- readDNAStringSet(filepath = paste("C:/Users/Guillem/Documents/PhD/comput/Sequences/newSeqs/Archive",
                                                    allSeqs[i], sep = "/"))
        allGenomes[[i]] <- genome
}
allSeqs <- gsub(".fasta|.fna", "", allSeqs)
names(allGenomes) <- allSeqs

# Retrieve compounds that are in the chosen pathways
cysMetMets <- gsub("\\..*", "", loadsPaths$metabolite[loadsPaths$Pathway == "Cysteine and methionine metabolism"])
cysMetMetsKEGGIDS <- dictionary$`KEGG IDs`[match(cysMetMets, dictionary$definitiveNames)]

TCAMets <- gsub("\\..*", "", loadsPaths$metabolite[loadsPaths$Pathway == "Citrate cycle (TCA cycle)"])
TCAMetsKEGGIDs <- dictionary$`KEGG IDs`[match(TCAMets, dictionary$definitiveNames)]



# Retrieve the enzymes that appear in these pathways
cysMetEnzsPerMet <- getEnzsRelWmetsInPath(cysMetMetsKEGGIDS, "map00270")
TCAEnzsPerMet <- getEnzsRelWmetsInPath(TCAMetsKEGGIDs, "map00020")

dictEnzymes[match(cysMetEnzsPerMet[[7]], dictEnzymes$ECnums), ]

# keep only the enzymes that are core across all strains
cysMetCoreEnzs <- getCoreEnzsXMet(cysMetEnzsPerMet)
TCACoreEnzs <- getCoreEnzsXMet(TCAEnzsPerMet)

# set train control for doing random forests 
#trControl <- trainControl(method = "cv",
#                          number = nCV,
#                          search = "grid",
#                          allowParallel = T)

# Do rForest of glycine dehydrogenase genes Chen found that were significateive in protein, bt if W70332 is
# removed this signifficance disappears. W70332 has a tail generated by rast annotation that in prokka annotation
# is not present

ubiGenes <- names(ParsedSubGraphs_allStrains_named_new)[grep("ubi", names(ParsedSubGraphs_allStrains_named_new))]
coqGenes <- names(ParsedSubGraphs_allStrains_named_new)[grep("COQ", names(ParsedSubGraphs_allStrains_named_new))]
nqrGenes <- names(ParsedSubGraphs_allStrains_named_new)[grep("nqr", names(ParsedSubGraphs_allStrains_named_new))]
nuoGenes <- names(ParsedSubGraphs_allStrains_named_new)[grep("nuo", names(ParsedSubGraphs_allStrains_named_new))]
sdhGenes <- names(ParsedSubGraphs_allStrains_named_new)[grep("sdh", names(ParsedSubGraphs_allStrains_named_new))]

gcvP_2_rForest <- doMultRForests("gcvP_2")

gcvP_1_rForest <- doMultRForests("gcvP_1")

gcvP_genes_rForest <- doMultRForests(c("gcvP_1", "gcvP_2"))

fabG_genes_rForest <- doMultRForests(c("fabG_11",
                                       "fabG_9",
                                       "fabG_4",
                                       "fabG_2",
                                       "fabG_10",
                                       "fabG_3",
                                       "fabG_5",
                                       "fabG_6",
                                       "fabG_8"))

ubiGenes_rForest <- doMultRForests(ubiGenes)
coqGenes_rForest <- doMultRForests(coqGenes)
nqrGenes_rForest <- doMultRForests(nqrGenes)
nuoGenes_rForest <- doMultRForests(nuoGenes)
sdhGenes_rForest <- doMultRForests(sdhGenes)

fabG_2_AAAlign <- ParsedSubGraphs_allStrains_named_new$fabG_2$AAAlignment

names(fabG_2_AAAlign) <- allSeqs

fabG_2_AA_SNPsMat <- buildSNPsMat(fabG_2_AAAlign, rhamnMat)

fabG_2_AA_SNPsMatBin <- binaryCodeSNPs(fabG_2_AA_SNPsMat)

fabG_2_SNPsCountVec <- rowSums(apply(fabG_2_AA_SNPsMatBin[, c(1, 3, 6, 7, 10)], 2, as.numeric))

names(fabG_2_SNPsCountVec) <- rownames(fabG_2_AA_SNPsMat)

wilcox.test(fabG_2_SNPsCountVec[fabG_2_AA_SNPsMat$rhamn2cats == 0],
            fabG_2_SNPsCountVec[fabG_2_AA_SNPsMat$rhamn2cats == 1])

trControl <- trainControl(method = "cv",
                          number = 60,
                          search = "grid",
                          allowParallel = T)

fabG_2_AA_fit <- train(rhamn2cats ~ . ,
                       data = fabG_2_AA_SNPsMat[, -(ncol(fabG_2_AA_SNPsMat))],
                       method = "rf",
                       metric = "Accuracy",
                       trControl = trControl)

predict(fabG_2_AA_fit, fabG_2_AA_SNPsMat[, -(ncol(fabG_2_AA_SNPsMat))])

doRForest(fabG_2_AA_SNPsMat)



# group_9868 is a hypothetical protein symilar to rtcB that was found to be related with oxidative stress. This gene
# controls the expression of the operon mexXY operon. This operon controls the expression of the multidrug efflux 
# system, which is upregulated in oxidative stress conditions. The RNAseq data showed that these genes are upregulated in
# non rhamnolipid producers (and non swarmers)

group_9868_rForests <- doMultRForests("group_9868")

rtcB_rForests <- doMultRForests(c("rtcB_1", "rtcB_2"))

rhlA_1_rForest <- doMultRForests("rhlE_3")

# Genes related with rhamnolipid production
rhl_genes_rForest <- doMultRForests(c("rhlA_1", "rhlA_2", "rhlB", "rhlE_1", "rhlE_2", "rhlE_3"))


rhlA_PAO1 <- "ATGCGGCGCGAAAGTCTGTTGGTAACGGTATGCAAGGGCCTGCGGGTACATGTCGAGCGCGTGGGGCAGGATCCCGGGCGCGACACGGTGATGCTGGTCAACGGCGCGATGGCGACCACCGCCTCGTTCGCCCGGACCTGCAAGTGCCTGGCCGAACATTTCAACGTGGTGCTGTTCGACCTGCCCTTCGCCGGGCAGTCGCGGCAGCACAATCCGCAGCGCGGGTTGATCACCAAGGACGACGAGGTGGAGATTCTCCTGGCGCTGATCGAGCGCTTCGCTGTCAACCACCTGGTCTCGGCCTCCTGGGGCGGCATCTCCACGCTGCTGGCGCTGTCGCGCAACCCGCGCGGGGTCCGCAGCTCGGTGGTGATGGCGTTCGCGCCGGGGCTGAACCAGGCGATGCTCGATTATGTCGGGCGGGCCCAGGAACTGATCGAACTGGACGACAAGTCGGCGATCGGCCACCTGCTCAACGAGACCGTCGGCAAGTACCTGCCGCCGCGGCTGAAGGCCAGCAACCATCAGCACATGGCCTCCCTGGCCACTGGCGAGTACGAGCAGGCGCGTTTCCACATCGACCAGGTGCTGGCGCTCAATGACCGTGGCTACCTGAGCTGCCTGGGGCAGATCCAGAGTCACGTGCATTTCATCAACGGCAGCTGGGACGAGTACACCACCGCCGAGGACGCCCGCCAGTTCCGCGATTACCTGCCGCATTGCAGTTTTTCGCGGGTGGAAGGCACCGGGCACTTCCTCGACCTGGAGTCCAAGCTGGCGGCGGCGCGTGTGCACCGGGCGTTGCTCGAGCACCTGCTGGCGCAACCGGAACCGTGGCGCTCCGAGCAGGCGGCGGGATTCCACGAGATGGCCATCGGCTACGCCTGA"

rhlA_1_wPAO1 <- c(sapply(ParsedSubGraphs_allStrains_named_new$rhlA_1$DNAAlignment, function(x) as.character(x)), 
                  rhlA_PAO1)

rhlA_1_wPAO1 <- DNAStringSet(rhlA_1_wPAO1)


BrowseSeqs(AlignTranslation(rhlA_1_wPAO1))

BrowseSeqs(AlignTranslation(rhlA_1_wPAO1, type = "AAStringSet"))
# folD gene participates in the synthesis of formyl group

folD_rForest <- doMultRForests("folD")

doMultRForests(c("sdhA", "sdhB"))

cysMetClustsAndRandForests <- retrAlignAndDends(cysMetCoreEnzs)
TCAClustsAndRandForests <- retrAlignAndDends(TCACoreEnzs)



# We're gonna do the randomForest with all aligned core genes
coreGenePos <- sapply(ParsedSubGraphs_allStrains_named_new, function(x) length(x$AAAlignment) == 30)

coreHeronAlignedGenes <- names(ParsedSubGraphs_allStrains_named_new)[coreGenePos]

doMultRForests(coreHeronAlignedGenes[400:405], maxWaitTime = 30, nCV = 5)

doMultRForests(coreHeronAlignedGenes[400])

# as there are many 4541 core aligned genes and it takes too much to run. Split the gene set to don't lose work
count <-  0
i <- 1
coreAligned_rForestList_1 <- list()
while (count < 226){
        coreAligned_rForest_subSet <- doMultRForests(coreHeronAlignedGenes[(count+1):(count+45)], maxWaitTime = 10, nCV = 5)
        coreAligned_rForestList_1[[i]] <- coreAligned_rForest_subSet
        count <- count + 45
        i <- i + 1
}



coreHeronAlignedGenes_rForests <- doMultRForests(coreHeronAlignedGenes, maxWaitTime = 15, nCV = 5)

doMultRForests(coreHeronAlignedGenes[1:5], maxWaitTime = 10, nCV = 5)

save(coreHeronAlignedGenes_rForests, file = "coreHeronAlignedGenes_rForests.RData")




g1979_AAAlign <- ParsedSubGraphs_allStrains_named_new$group_1979$AAAlignment

names(g1979_AAAlign) <- allSeqs

g1979_SNPs <- buildSNPsMat(g1979_AAAlign, rhamnMat)

asdfg <- sapply(ParsedSubGraphs_allStrains_named_new, function(x) ncol(buildSNPsMat(x$AAAlign, rhamnMat))-2 == 2)

asdfg <- asdfg[sapply(asdfg, function(x) length(x) != 0)]


asdfgTrue <- names(asdfg)[unlist(asdfg)]
snps2cols <- list()
for(i in seq_along(asdfgTrue)){
        snps2cols[[i]] <- buildSNPsMat(ParsedSubGraphs_allStrains_named_new[[match(asdfgTrue[i],
                                                                                  names(ParsedSubGraphs_allStrains_named_new))]]$AAAlign,
                                       rhamnMat)
}
names(snps2cols) <- asdfgTrue

# This gene has 2 snps that are only present in one strain
doMultRForests("group_4558")

group_4558 <- ParsedSubGraphs_allStrains_named_new$group_4558$AAAlignment
names(group_4558) <- allSeqs[as.numeric(gsub("\\_.*", "", names(group_4558)))]

g4558_SNPs <- buildSNPsMat(group_4558, rhamnMat)
g4558_SNPs_tab <- table(g4558_SNPs[, 1:2])

min(colSums(g4558_SNPs_tab)) == 1
min(rowSums(g4558_SNPs_tab)) == 1

pos2Code <- names(sort(colSums(g4558_SNPs_tab)))[1]

match(pos2Code, g4558_SNPs[, 2])

sum(g4558_SNPs_tab == 0)


which(sapply(snps2cols, function(x) sum(table(x[, 1:2])==0) ==2))




doRForest(g1979_SNPs)

# Set time control for avoiding r getting stuck in problematic sequences



trControl <- trainControl(method = "cv",
                          number = 60,
                          search = "grid",
                          allowParallel = T)


model <- R.utils::withTimeout({
        caret::train(rhamn2cats ~ ., 
        data = g4558_SNPs[, 1:(ncol(g4558_SNPs)-1)],
        method = "rf",
        trControl = trControl, 
        metric = "Accuracy")
}, timeout = 5, 
onTimeout = "silent")

model <- R.utils::withTimeout({
        doRForest(g4558_SNPs)}, timeout = 5, 
onTimeout = "silent")




withTimeout(
        theModel <- train(rhamn2cats ~ .,
                          data = g4558_SNPs[, 1:(ncol(g4558_SNPs)-1)],
                          method = "rf",
                          metric = "Accuracy",
                          trControl = trControl),
        timeout=1, 
        onTimeout = "silent")


foo1 <- function(n = 1000000) { 
        ret <- rep(0, n); 
        for (kk in 1:n) ret[kk] <- rnorm(1); 
        ret; 
}

g4558_SNPs

# The following will time out after 2s
tryCatch( { res <- withTimeout( { doRForest(g4558_SNPs) },
                                timeout = 2) },
          TimeoutException = function(ex) cat("Timed out\n"))





ParsedSubGraphs_allStrains_named_new$dnaA$AAAlignment

unlist(cysMetCoreEnzs) %in% names(ParsedSubGraphs_allStrains_named_new)
unlist(TCACoreEnzs) %in% names(ParsedSubGraphs_allStrains_named_new)

a <- ParsedSubGraphs_allStrains_named_new[names(ParsedSubGraphs_allStrains_named_new) == unlist(cysMetCoreEnzs)[1]][[1]]

fmtAAAlign <- ParsedSubGraphs_allStrains_named_new$fmt$AAAlignment
names(fmtAAAlign) <- allSeqs[as.numeric(gsub("\\_.*", "", names(fmtAAAlign)))]






# FMT aligmnment 

fmtAAdist <- ParsedSubGraphs_allStrains_named_new$fmt$AADist
colnames(fmtAAdist) <- allSeqs
rownames(fmtAAdist) <- allSeqs

fmtAAdist <- as.dist(fmtAAdist)

fmtAAdist_hca <- IdClusters(myDistMatrix = fmtAAdist, myXStringSet = fmtAAAlign, method = "complete",
                            type = "dendrogram")

dend <- ggdendrogram(fmtAAdist_hca)




ParsedSubGraphs_allStrains_named_new$fmt$AADend


fmtDendDat <- dendro_data(fmtAAdist_hca)



fmtLabs <- label(fmtDendDat)

rhamn <- rhamnMat$rhamn3cats[match(fmtLabs$label, rhamnMat$strains)]

fmtLabs$rhamn <- as.factor(rhamn)


dend <- ggplot(segment(fmtDendDat)) +
        geom_segment(aes(x=x, y=y, xend=xend, yend=yend))

dend + geom_text(data=label(fmtDendDat),
                 aes(label=label, x=x, y=-0.0001, colour=fmtLabs$rhamn, angle = 90, hjust = "top"),
                 nudge_y = -0.0001) +
        ylim(-0.001, 0.008) + 
        scale_color_manual(values = c("blue", "green","red"))
ggsave("FMTAAAlignDend.pdf")




fmtSNPsMat <- buildSNPsMat(fmtAAAlign, rhamnMat)

fmtSNPsMat_bin <- binaryCodeSNPs(fmtSNPsMat)

if(!require(randomForest)) install.packages("randomForest")

fmtRForest <- doRForest(fmtSNPsMat_bin)


doRForest(binaryCodeSNPs(buildSNPsMat(ParsedSubGraphs_allStrains_named_new$rhlA_1$AAAlignment, rhamnMat)))

# rhl enzymes 

rhlA_1_AAAlig <- ParsedSubGraphs_allStrains_named_new$rhlA_1$AAAlignment

names(rhlA_1_AAAlig) <- allSeqs

rhlA_1_SNPs <- buildSNPsMat(rhlA_1_AAAlig, rhamnMat)


rhlA_1_rf <- train(rhamn2cats ~ .,
                   data = rhlA_1_SNPs[, 1:(ncol(rhlA_1_SNPs)-1)],
                   method = "rf",
                   metric = "Accuracy",
                   trControl = trControl)

rhlA_1 <- retrAlignAndDends("rhlA_1")
rhlA_2 <- retrAlignAndDends("rhlA_2")
rhl_B <- retrAlignAndDends("rhl_B")
rhlE_1 <- retrAlignAndDends("rhlE_1")
rhlE_2 <- retrAlignAndDends("rhlE_2")
rhlE_3 <- retrAlignAndDends("rhlE_3")

dictEnzymes[dictEnzymes$Gene %in% c("rhlA_1", "rhlA_2", "rhl_B", "rhlE_1", "rhlE_2", "rhlE_3"),]




lst <- lapply(geneCallsAllStrains4HeronProkka,
              function(x) x[grep("gcvP_1", x$Annotation), ])
lst <- lst[sapply(lst, function(x) nrow(x) > 0)]
seqVec <- c()
for(i in seq_along(lst)){
        if(lst[[i]]$Strand == 0){
                seq <- as.character(allGenomes[[grep(names(lst)[i],
                                                     names(allGenomes))]][[1]][lst[[i]]$Start:lst[[i]]$Stop])
        }else{
                seq <- allGenomes[[grep(names(lst)[i],
                                        names(allGenomes))]][[1]][lst[[i]]$Stop:lst[[i]]$Start]
                seq <- complement(seq)
                seq <- as.character(seq)
                #seq <- paste(rev(strsplit(seq, NULL)[[1]]), collapse = "")
        }
        seqVec <- c(seqVec, seq)
}
strSet <- DNAStringSet(seqVec)
DNA_alig <- AlignTranslation(strSet, type = "DNAStringSet")
AA_alig <- AlignTranslation(strSet, type = "AAStringSet")

BrowseSeqs(DNA_alig)
lst[29]

lst_2 <- lapply(geneCallsAllStrains4HeronProkka,
                function(x) x[grep("gcvP_2", x$Annotation), ])
lst_2
# Glycine dehidrogenase (Chen found that could predict rhamn production with it)
gcvP_2Align <- doAlignFromGeneCalls("gcvP_2")
gcvP_1Align <- doAlignFromGeneCalls("gcvP_1")
BrowseSeqs(gcvP_1Align$DNA)
BrowseSeqs(gcvP_1Align$AA)




# The annotation of gcvP_1 (gcvP_1 is the enzyme that chen correlated with rhamn production) is not correct. It was indeed correct. I am retarded and
# was not aligning the complementary seq in negative dna strands.
gffsAllStrains$F22031[which(gffsAllStrains$F22031$locus_tag == "JJIFMKBL_02725"),]

gffsAllStrains$F22031[which(gffsAllStrains$M1608$locus_tag == "EIIJMEBM_03216"),]
gffsAllStrains$M37351[which(gffsAllStrains$M37351$locus_tag == "INGCDNNM_03661"),]
gffsAllStrains$M55212[which(gffsAllStrains$M55212$locus_tag == "PEDOHHHI_05192"),]

gcvP_1SNPsMat <- buildSNPsMat(gcvP_1Align$AA, rhamnMat)
gcvP_2SNPsMat <- buildSNPsMat(gcvP_2Align$AA, rhamnMat)

doRForest(gcvP_1SNPsMat)
doRForest(gcvP_2SNPsMat)


gcvP_2SNPsMat_bin <- binaryCodeSNPs(gcvP_2SNPsMat)
gcvP_1SNPsMat_bin <- binaryCodeSNPs(gcvP_1SNPsMat)

gcvP_1SNPsMat_bin_rForest <- doRForest(gcvP_1SNPsMat_bin)
gcvP_2SNPsMat_bin_rForest <- doRForest(gcvP_2SNPsMat_bin)

gcvP_2SNPsMat_bin_rForest$`3cats`


BrowseSeqs(gcvP_2Align$AA)
